Google.Api.Gax.Grpc/ClientBuilderBase.cs

```diff
@@ -123,6 +143,93 @@ protected virtual void Validate()
+        /// <returns>A key/value mapping of the emulator environment variables to their values, or null if the emulator should not be used.</returns>
+        protected Dictionary<string, string> GetEmulatorEnvironment(
+            IEnumerable<string> requiredEmulatorEnvironmentVariables,
+            IEnumerable<string> allEmulatorEnvironmentVariables)
+        {
+            var environment = allEmulatorEnvironmentVariables.ToDictionary(GetEnvironmentVariableOrNull);
+            
+            switch (EmulatorDetection)
+            {
+                case EmulatorDetection.None:
+                    return default;
+                case EmulatorDetection.ProductionOnly:
+                    foreach (var variable in allEmulatorEnvironmentVariables)
+                    {
+                        GaxPreconditions.CheckState(
+                            environment[variable] is null,
+                            "Emulator environment variable '{0}' is set, contrary to use of {1}.{2}",
+                            variable, nameof(EmulatorDetection), nameof(EmulatorDetection.ProductionOnly));
+                    }
+                    return default;
+                case EmulatorDetection.EmulatorOnly:
+                    foreach (var variable in requiredEmulatorEnvironmentVariables)
+                    {
+                        GaxPreconditions.CheckState(
+                            environment[variable] is object,
+                            "Emulator environment variable '{0}' is not set, contrary to use of {1}.{2}",
+                            variable, nameof(EmulatorDetection), nameof(EmulatorDetection.EmulatorOnly));
+                    }
+                    break;
+                case EmulatorDetection.EmulatorOrProduction:
+                    bool anySet = allEmulatorEnvironmentVariables.Any(v => environment[v] is object);
+                    if (!anySet)
+                    {
+                        return default;
+                    }
+                    bool allRequiredSet = requiredEmulatorEnvironmentVariables.All(v => environment[v] is object);
+                    if (!allRequiredSet)
+                    {
+                        var sampleSet = allEmulatorEnvironmentVariables.First(v => environment[v] is object);
+                        var sampleNotSet = requiredEmulatorEnvironmentVariables.First(v => environment[v] is object);
+                        GaxPreconditions.CheckState(false,
+                            "Emulator environment variable '{0}' is set, but '{1}' is not set.",
+                            sampleSet, sampleNotSet);
+                    }
+                    break;
+                default:
+                    throw new InvalidOperationException($"Invalid emulator detection value: {EmulatorDetection}");
+            }
+
+            // We know we want to use the emulator. Validate nothing else has been set that shouldn't be.
```

Comment:
We disagreed on this in the previous prototype too :) I think it's more dangerous to have one setting that means "ignore everything else silently".
Definitely up for further discussion of this, but I think it can be separate to the refactoring part.